server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java

```diff
@@ -142,11 +147,32 @@ public void clusterStateProcessed(final String source, final ClusterState oldSta
                         assert blockedIndices.isEmpty() == false : "List of blocked indices is empty but cluster state was changed";
                         threadPool.executor(ThreadPool.Names.MANAGEMENT)
                             .execute(new WaitForClosedBlocksApplied(blockedIndices, timeout,
-                                ActionListener.wrap(closedBlocksResults ->
+                                ActionListener.wrap(results ->
                                     clusterService.submitStateUpdateTask("close-indices", new ClusterStateUpdateTask(Priority.URGENT) {
+
+                                        boolean acknowledged = true;
+
                                         @Override
                                         public ClusterState execute(final ClusterState currentState) throws Exception {
-                                            final ClusterState updatedState = closeRoutingTable(currentState, closedBlocksResults);
+                                            final ClusterState updatedState = closeRoutingTable(currentState, blockedIndices, results);
+                                            // Combine the results of the verify shards before close actions with the cluster state changes
+                                            // to determine if the current close action effectively closed all indices.
+                                            for (Map.Entry<Index, AcknowledgedResponse> result : results.entrySet()) {
+                                                IndexMetaData updatedMetaData = updatedState.metaData().index(result.getKey());
+                                                if (updatedMetaData != null) {
+                                                    if (result.getValue().isAcknowledged()) {
+                                                        if (updatedMetaData.getState() == IndexMetaData.State.CLOSE) {
+                                                            IndexMetaData previousMetaData = currentState.metaData().index(result.getKey());
+                                                            if (previousMetaData != null) {
+                                                                acknowledged = (previousMetaData.getState() == IndexMetaData.State.OPEN);
```

Comment:
I'm not sure I follow this logic here. Wouldn't it be much simpler just to say. If the index is closed at this point, all is good. If not, something must have gone wrong, thereby acked == false.

This amounts to the following simple logic:

```
for (Map.Entry<Index, AcknowledgedResponse> result : results.entrySet()) {
    IndexMetaData updatedMetaData = updatedState.metaData().index(result.getKey());
    if (updatedMetaData != null && updatedMetaData.getState() != IndexMetaData.State.CLOSE) {
        acknowledged = false;
        break;
    }
}
```

Reply:
Agreed, let's do that.